src/main/ruby/truffleruby/core/file.rb

```diff
@@ -415,37 +415,26 @@ def self.last_nonslash(path, start=nil)
   # the separator used on the local file system.
   #
   #  File.dirname("/home/gumby/work/ruby.rb")   #=> "/home/gumby/work"
-  def self.dirname(path)
-    path = Truffle::Type.coerce_to_path(path)
-
-    # edge case
-    return +'.' if path.empty?
-
-    slash = '/'
-
-    # pull off any /'s at the end to ignore
-    chunk_size = last_nonslash(path)
-    return +'/' unless chunk_size
-
-    if pos = Primitive.find_string_reverse(path, slash, chunk_size)
-      return +'/' if pos == 0
-
-      path = path.byteslice(0, pos)
-
-      return +'/' if path == '/'
-
-      return path unless path.end_with? slash
+  def self.dirname(path,num_levels=1)
+    if num_levels < 0
+      raise ArgumentError, "level can't be negative"
+    end
+    #This must happen before the type coercion to string below, because File.dirname accepts any objects which can respond to a :to_path message
+    #And if path was such an object, then a File.dirname(obj,0) returns it as-is without conversion to string
+    #Not that intuitive, but that's how Matz ruby behave
+    if num_levels == 0
+      return path
+    end
 
-      # prune any trailing /'s
-      idx = last_nonslash(path, pos)
+    path = Truffle::Type.coerce_to_path(path)
 
-      # edge case, only /'s, return /
-      return +'/' unless idx
+    num_levels.times do
+      # edge case
+      return +'.' if path.empty?
```

Comment:
Does it make sense to check `path.empty?` on each iteration but not only once at the begining?

Reply:
I guess so, since I'm modifying path every iteration, it can potentially become empty every iteration. 

This check is actually the first check in the original File.dirname, I just left it in its place here instead of moving it with the rest into FileOperations.dirname_once, because it seems intuitive that if a path ever becomes empty that we early return right away instead of continuing to call dirname_once uselessly.

Reply:
Not a big deal actually, but the helper method returns either:
- "."
- "/"
- `path.byteslice(0, pos)`
- `path.byteslice(0, idx - 1)`

So I assume it cannot return empty String.

Reply:
I tested it just now in ruby 3.1.0, it can : 

irb(main):004:0>  "hello".byteslice(0,0)
=> ""

so if idx is 1 or pos is 0 it will return an empty string.